Add support for Windows Secure Hot-Patching

sivadeilra opened the pull request:
This PR adds some of the support needed for Windows hot-patching.

Windows implements a form of hot-patching. This allows patches to be applied to Windows apps, drivers, and the kernel, without rebooting or restarting any of these components. Hot-patching is a complex technology and requires coordination between the OS, compilers, linkers, and additional tools.

This PR adds support to Clang and LLVM for part of the hot-patching process.  It enables LLVM to generate the required code changes and to generate CodeView symbols which identify hot-patched functions.  The PR provides new command-line arguments to Clang which allow developers to identify the list of functions that need to be hot-patched.  This PR also allows LLVM to directly receive the list of functions to be modified, so that language front-ends which have not yet been modified (such as Rust) can still make use of hot-patching.

This PR:

* Adds a `MarkedForWindowsHotPatching` LLVM function attribute. This attribute indicates that a function should be _hot-patched_. This generates a new CodeView symbol, `S_HOTPATCHFUNC`, which identifies any function that has been hot-patched. This attribute also causes accesses to global variables to be indirected through a `_ref_*` global variable. This allows hot-patched functions to access the correct version of a global variable; the hot-patched code needs to access the variable in the _original_ image, not the patch image.
* Adds a `AllowDirectAccessInHotPatchFunction` LLVM attribute. This attribute may be placed on global variable declarations. It indicates that the variable may be safely accessed without the `_ref_*` indirection.
* Adds two Clang command-line parameters: `-fms-hotpatch-functions-file` and `-fms-hotpatch-functions-list`.  The `-file` flag may point to a text file, which contains a list of functions to be hot-patched (one function name per line).  The `-list` flag simply directly identifies functions to be patched, using a comma-separated list.  These two command-line parameters may also be combined; the final set of functions to be hot-patched is the union of the two sets.
* Adds similar LLVM command-line parameters: `--ms-hotpatch-functions-file` and `--ms-hotpatch-functions-list`.
* Adds integration tests for both LLVM and Clang.
* Adds support for dumping the new `S_HOTPATCHFUNC` CodeView symbol.

Although the flags are redundant between Clang and LLVM, this allows additional languages (such as Rust) to take advantage of hot-patching support before they have been modified to generate the required attributes.

Credit to @dpaoliello, who wrote the original form of this patch.

sivadeilra committed:
Windows hotpatching support

move hotpatch tests to X86 subdir

move llvm hotpatch tests to X86 dir

switch to strbool attributes

cleanup

require x86 target for tests

sivadeilra commented:
For context, here are recent Microsoft announcements concerning hot-patching:

* https://www.microsoft.com/en-us/windows-server/blog/2025/04/24/tired-of-all-the-restarts-get-hotpatching-for-windows-server/?msockid=19a6f8f09bd160ac0b18ed449afc614b
* https://techcommunity.microsoft.com/blog/windows-itpro-blog/hotpatch-for-windows-client-now-available/4399808

efriedma-quic commented on llvm/lib/CodeGen/WindowsHotPatch.cpp:
```diff
@@ -0,0 +1,251 @@
+        RefMapping.try_emplace(GVUse.GV).first->second;
+    if (ReplaceWithRefGV == nullptr) {
+      Constant *AddrOfOldGV = ConstantExpr::getGetElementPtr(
+          Builder.getPtrTy(), GVUse.GV, ArrayRef<Value *>{});
```
The getGetElementPtr call here just returns GVUse.GV, I think?  We do constant folding on constant expressions.

efriedma-quic commented on llvm/lib/CodeGen/WindowsHotPatch.cpp:
```diff
@@ -0,0 +1,251 @@
+    SmallDenseMap<GlobalVariable *, GlobalVariable *> &RefMapping,
+    DIBuilder &DebugInfo) {
+  for (auto &GVUse : GVUses) {
+    IRBuilder<> Builder(GVUse.User);
```
You can't insert code before an arbitrary instruction. For example, a PHI node.

sivadeilra replied:
I see.  What would you suggest?  What IR could I build that would repro the problem?  Would a PHI node directly consume a GlobalVariable node?

efriedma-quic replied:
Something like the following gives a PHI with a global as an operand:

```
int g1, g2;
int* f(int b, void g()) {
  int* ret;
  if (b) {
    g();
    ret = &g1;
  } else {
    ret = &g2;
  }
  return ret;
}
```

For PHI nodes, you can just insert the load into the predecessor BB, or something like that.  Well, unless there's a catchswitch involved.

Most other instructions won't be a problem here... some less common edge cases to consider: catchpad, llvm.threadlocal.address.

efriedma-quic replied:
Inserting in the entry block should be fine in terms of correctness.  If the function is large, the address is likely to be spilled to the stack, though.

sivadeilra commented on llvm/lib/CodeGen/WindowsHotPatch.cpp:
```diff
@@ -0,0 +1,251 @@
+    // marked for hotpatching using the list provided directly to LLVM.
+    for (auto &F : M.functions()) {
+      // Ignore declarations that are not definitions.
+      if (F.isDeclarationForLinker()) {
```
No, I think we want to include them.  Current hot-patch tools cannot patch symbols that are marked as private, but that is a limitation of the hot-patch tools.  Correlating private symbols in before/after builds of the same module is feasible and desirable, so I don't see any reason to exclude them from hot-patching here.

efriedma-quic replied:
The primary problem here is that optimization passes can manipulate internal function symbols.  What do you do if you want to hotpatch a symbol, but it's disappeared, or has a different signature?

sivadeilra replied:
That's a problem we already have with MSVC.  Our hot-patch tooling doesn't make guarantees, and developers are currently required to do a moderate amount of "homework" while preparing hot-patches.  They already have to deal with a function getting inlined into callers (which is why the `S_INLINEES` record was added, months ago) and several other situations.

Right now, we're aiming for parity with the experience using MSVC.  Once we reach that, we can work on doing better than MSVC.  Getting to parity will be no small achievement in the first place.

Developers already have to avoid changing function signatures, for example, for any call that crosses the patched / not-patched boundary.

MolecularMatters replied:
I think I don't follow.
There are hot-patch tools that can deal with symbols with internal linkage just fine ([Live++](https://liveplusplus.tech)), and it doesn't require developers to do any "homework" at all.

Furthermore, you mentioned the experience using MSVC - are you referring to Edit & Continue or something else?

sivadeilra replied:
"We" in this context is Microsoft, specifically the Windows organization.  We have in-house tools that generate our hot-patches.  Currently, these tools only work with MSVC.  The aim of this PR (and others, in the future) is to allow these tools to work with LLVM, specifically with Clang and Rust.

Our tools currently do not handle static functions.  However, I don't want to exclude the possibility of supporting static functions in the future, which is why I responded to @efriedma-quic 's comment by saying that I don't want to restrict this code to handle only public functions.

dpaoliello replied:
> Furthermore, you mentioned the experience using MSVC - are you referring to Edit & Continue or something else?

This is not for Edit-and-Continue, this is for hot-patching deployed Windows machines running retail bits: https://www.microsoft.com/en-us/windows-server/blog/2025/04/24/tired-of-all-the-restarts-get-hotpatching-for-windows-server/?msockid=19a6f8f09bd160ac0b18ed449afc614b

> There are hot-patch tools that can deal with symbols with internal linkage just fine ([Live++](https://liveplusplus.tech/)), and it doesn't require developers to do any "homework" at all.

True, but tools like that have their own restrictions. From their own docs:
>Linker -> Optimization -> References must be set to No (/OPT:NOREF)
>Linker -> Optimization -> Enable COMDAT Folding must be set to No (/OPT:NOICF)

For Windows hot-patching one of the goals was to enable its use with fully-optimized binaries, the downside of this (as Eli called out) that that the optimizer can do lots of things that mess with the resulting symbols and binaries, hence why there is some "homework" required (i.e., we'd rather have Microsoft devs spend time to craft hot-patching changes that meet these requirements rather than deploy suboptimal binaries to customers).

sivadeilra commented on llvm/lib/CodeGen/WindowsHotPatch.cpp:
```diff
@@ -0,0 +1,251 @@
+      }
+
+      llvm::StringRef FuncName{};
+      if (auto *SP = F.getSubprogram(); SP != nullptr) {
```
Ahh, I think this is incorrect.  I just need the mangled name of the symbol, and at one point this was the only way I could get it.  I think this can simply be `F.getName()`; I'll verify and update.

sivadeilra replied:
Updated to use simply `F.getName().

aganea commented:
Since the [MSVC flag](https://learn.microsoft.com/en-us/cpp/build/reference/hotpatch-create-hotpatchable-image?view=msvc-170) `/HOTPATCH` has been supported in LLVM and clang-cl for quite a while now (see [this](https://reviews.llvm.org/D116511), and #77245 and #87639) the description in this PR is quite confusing. Or maybe is it the Microsoft nomenclature for this user-facing feature which can be confused with what used to be called hotpatching before?

I'd like to understand the history here. It seems `/HOTPATCH` and `/FUNCTIONPADMIN` were initially added to MSVC for this exact purpose during the Windows NT era (probably around 2010?) but have never been used for this purpose. It looks like this new "hotpatching" described in the links is a revival of that idea?

Can you provide a more technical insight about how is this gonna be used? What are the equivalent MSVC flags? Why is it that users (programmers) need to select specific functions to hotpatch?

As a side-note, we (at Sony) have been using [Live++](https://liveplusplus.tech/) for this purpose for quite a while now on large game projects, sometimes going for a day long, on hundered of iterations, without stopping the game. This scenario (patching global variables) is already well supported by Live++ out of the box without the changes proposed in this PR. However I understand that Microsoft has a different tech that requires these changes.

Also, this paragraph is a bit confusing:

> Adds a MarkedForWindowsHotPatching LLVM function attribute. This attribute indicates that a function should be hot-patched. This generates a new CodeView symbol, S_HOTPATCHFUNC, which identifies any function **that has been hot-patched**. This attribute also causes accesses to global variables to be indirected through a _ref_* global variable. This allows hot-patched functions to access the correct version of a global variable; the hot-patched code needs to access the variable in the original image, not the patch image.

"hot-patching/hot-patched" usually refers to a process that happens at runtime, not at compile time. Can you explain more in (technical) detail the whole pipeline for this new hotpatch feature? What does exactly the kernel/the system do after a new Windows update is downloaded? Does it iterate on the debug infos for each DLL in the update package, and if it finds `S_HOTPATCHFUNC` it would patch/replace the functions in memory with the new DLL? (DLL which I expect, is previously injected in the process). Why do we need to redirect the _ref_* global variables? Can't the kernel loader parse the DLL relocs and patch/redirect the injected DLL functions to the initial global state?

+ @MolecularMatters for visibility

sivadeilra commented on llvm/include/llvm/DebugInfo/CodeView/SymbolRecord.h:
```diff
@@ -176,6 +176,21 @@ class CallerSym : public SymbolRecord {
   uint32_t RecordOffset = 0;
 };
 
+class HotPatchFuncSym : public SymbolRecord {
```
Our hot-patch tools read PDBs and then generate metadata that is placed into the final DLL/SYS/EXE, into a COFF section that is reserved for this purpose.

aganea replied:
Thanks! Would that tooling remain proprietary, or is it something that will be open-sourced as well, for example for NT drivers developpers?

sivadeilra commented on clang/test/CodeGen/ms-hotpatch.c:
```diff
@@ -0,0 +1,20 @@
+// This verifies that hotpatch function attributes are correctly propagated when compiling directly to OBJ.
+//
+// RUN: %clang_cl -c --target=x86_64-windows-msvc -O2 /Z7 -fms-hotpatch-functions-file=%S/ms-hotpatch-functions.txt /Fo%t.obj %s
```
Thanks.  I've added checks for targeting Windows/COFF to the code.

aganea replied:
Thank you! Are you able to exclude (and warn for) 32-bit (i386) and ARM/ARM64? We have users that target these archs on Windows.

sivadeilra replied:
We do support hot-patching ARM64 (and ARM64EC / ARM64X), so I don't want to exclude those architectures.

Windows on 32-bit ARM is obsolete and hot-patching was never a target for it.  Even 32-bit x86 is, in theory, hot-patchable, even if the majority of our hot-patches target kernels (and therefore only apply to AMD64 and ARM64).

Is there a reason for checking specific architectures?  This feature is currently only usable by Microsoft engineers, because hot-patches must be signed and this is done only for MS CVE fixes.  There's no danger of enabling these options for other architectures; it won't cause any problems.

aganea replied:
It is simply to be clear about what is supported or not, if someone inadvertently tries to use these options on unsupported archs.

aganea commented:
Thanks for the answer.

Are these new flags `-fms-hotpatch-functions-file=` and `-fms-hotpatch-functions-list=` populated with data maintained manually by the Windows developpers?

If I understand correctly, what you describe as "symbols that were hot-patched" means functions that are ready/candidates for hotpatching at runtime on the OS that installs the hotpatch, is that right?

I get the part about the security and verification, but I still don't get the `_ref_` thing. Usually after injecting the patched DLLs/SYS files into a process, the patcher can (automatically, from the relocs) redirect pointers present in the injected DLL to actually use the global state that is being used by the in-memory DLL. Is this `_ref_` related to a synchronization thing? Or timing? I would assume the kernel still needs a sync point where it stops ticking the process, installs the patch, patches all the function pointers from the in-memory DLL to the new injected DLL?

It'd be really nice if someone at Microsoft could write at some point a supporting documentation to understand how all this is to be used (a tech blog or a white paper describing it?)

sivadeilra commented:
>  I still don't get the _ref_ thing. [...]

This is simply how Windows Secure Hotpatching works, and has for years.  It relies on different codegen for the hotpatched functions, rather than using fixups to modify code.  This way, code segments are not modified (so hashes and code integrity still work).  Ordinary fixups cannot be used (since they refer to relocations within a single image, not across multiple images (i.e. base image vs. patch image)), so using a different form of fixup would have been required.  Microsoft (MSVC + Windows) chose to implement this using the global indirection technique.

This is not directly related to synchronization, although if synchronization objects are stored in global variables it would certainly apply to them.

It might help to emphasize the scenario that this is meant to support: hot-patching CVEs (vulnerabilities) in system services, the OS kernel, and device drivers.  It's not meant for general development.

> I would assume the kernel still needs a sync point where it stops ticking the process, [...]

No, the target process is not stopped.  In fact, we apply hot-patches to the kernel itself, without halting any processes or CPUs.  The order of operations that Windows hot-patching uses guarantees the consistency requirements that we need.  One of those requirements is that all code sees the right data (always points into global variables in the original image, not the hot-patched image).

> It'd be really nice if someone at Microsoft could write at some point a supporting documentation to understand how all this is to be used (a tech blog or a white paper describing it?)

We're working in that direction.  This PR is one of our first steps in publicizing any part of the toolchain and workflow, which is the nature of the `S_HOTPATCHFUNC` symbol and the need for global variable indirection.  These two requirements are what affects the compiler.  I can't speak to when or whether the rest of the workflow could be made public.

aganea commented:
Oh so all this dance (`_ref_` and the additional metadata) is for code page integrity purposes? To keep them unmodified in memory? So how does then the kernel use the PE metadata if it doesn't patch the code memory pages of the initial (running) image? Is there an additional mechanism for trapping and redirecting calls into the new image? If that's the case, there's no really image patching involved at runtime? The "hot-patching" is just a _vue d'esprit_ to present the concept to the user?

sivadeilra commented:
> Oh so all this dance (`_ref_` and the additional metadata) is for code page integrity purposes only? To keep them unmodified in memory? So how does then the kernel use the PE metadata if it doesn't patch the code memory pages of the initial (running) image? Is there an additional mechanism for trapping and redirecting calls into the new image? If that's the case, there's no really image patching involved at runtime? The "hot-patching" part is just a _vue d'esprit_ to present the concept to the user?

Sorry, I may have unintentionally mislead.  Let me clarify.

Our workflow for generating patches is this:

1. A vulnerability is identified and the affected functions are identified.
2. Hot-patching requirements are checked (a combination of manual and automated checks).  It cannot introduce new DLL imports/exports, cannot change function signatures of existing functions (unless they are entirely inlined), cannot add new fields to existing types, etc.
3. An "intermediate patch image" is created. This is a normal compilation of a complete binary, but with the flags added to the compiler that tell it to hot-patch certain functions.  This step is the focus on this PR.  The compiler and linker generate a complete executable image, but the hot-patched functions use `__ref_*` indirection and the PDB contains a description of the hot-patched functions.
4. We run our hot-patch generation tool, which compares the original image (called the "base") with the "intermediate" image.  It automatically verifies many of our requirements; if those requirements fail, the developer must re-evaluate and go back to step 2.  The most common cause is that inlining caused a function to be pulled into the hot-patch set.  The output of this step is the "hot-patch metadata".
5. The tools modify a copy of the intermediate image and insert the "hot-patch metadata". This creates the "final image".  The final image can either be loaded as a complete, standalone binary (and will be, when the system reboots) or can be used to hot-patch an active instance of the base image.

The metadata describes code patches in _both_ directions, as well data patches in _one_ direction.  The code patches modify the base image in memory, using a code update idiom that does not require stopping threads or CPUs.  The code patches also modify the hot-patch image, so that function calls to non-patched code are modified to point back into the base image.  This is done so that multiple hot-patches of the same binary can occur; each new hot-patch completely replaces the code from the previous hot-patch, although existing threads (or CPUs) may continue executing those paths until they return from them.  The hot-patch metadata also describes how to set the initial value of the `__ref_*` variables to point into the variables in the base image.

So I didn't mean to imply that the sole purpose of the `__ref_*` pointers was to avoid modifying the original image, in memory.

The Windows kernel contains the code that interprets the hot-patch metadata, so the format and semantics of the hot-patch metadata are determined by that code.  This PR is meant to enable Clang (and eventually Rust) to generate code that can work in this workflow.

For reference, @dpaoliello and I are the authors of Rust code that executes in the Windows kernel.  Our motivation for this work is to enable hot-patching of this code and of related non-Rust code in the same images.  Aligning LLVM's codegen, in this situation, and providing the S_HOTPATCHFUNC symbol, are a necessary part of enabling this whole scenario and in continuing with Rust development within the Windows kernel.

majnemer commented on llvm/lib/CodeGen/WindowsHotPatch.cpp:
```diff
@@ -0,0 +1,251 @@
+  for (auto &I : instructions(F)) {
+    for (auto &U : I.operands()) {
+      // Discover all uses of GlobalVariable, these will need to be replaced.
+      GlobalVariable *GV = dyn_cast<GlobalVariable>(&U);
```
Probably something like:
```C++
SmallPtrSet<const Constant *, 8> Visited;
SmallVector<const Constant *, 8> WorkList;
...

if (const auto *C = dyn_cast<Constant>(&U)) {
  if (Visited.insert(C).second)
    WorkList.push_back(C);
 
  while (!WorkList.empty()) {
    const Constant *WorkItem = WorkList.pop_back_val();
    if (const auto *GV = dyn_cast<GlobalValue>(WorkItem)) {
      if (!GV->hasAttribute(Attribute::AllowDirectAccessInHotPatchFunction)) {
        unsigned OpNo = &U - I.op_begin();
        GVUses.push_back({GV, &I, OpNo});
      }
    }
    for (const Value *Op : WorkItem->operands()) {
      const Constant *ConstOp = dyn_cast<Constant>(Op);
      if (!ConstOp)
        continue;
      if (Visited.insert(ConstOp).second)
        WorkList.push_back(ConstOp);
    }
  }
}
```

zmodem commented:
This is very interesting!

Initially I also assumed this was about Edit&Continue. As users may have the same idea, I'd suggest adding `DocBrief`s to the new Clang options, and a comment at the top of llvm/lib/CodeGen/WindowsHotPatch.cpp that explain in more detail what they do.

sivadeilra commented:
> We already have /HOTPATCH or rather -fms-hotpatch that does something different

They do something different, but not unrelated.  The "Secure Hot-Patching" feature (the feature I'm working on) still relies on the function prolog using hot-patchable code generation.  So `/HOTPATCH` is also necessary for this, as well as the `/FUNCTIONPADMIN` linker option.

I agree that the terminology is a bit confusing, here, since "hot-patching" is a more general concept than the particular feature / workflow that I am working on, and there are many different implementations of the general idea of "hot-patching".  I can certainly add more explanatory text to `WindowsHotPatch.cpp` file.

What would you think of renaming the options to `-fms-secure-hotpatch-functions-file` and `-fms-secure-hotpatch-functions-list`?  The feature internally is known as "Secure Hot-Patching" (SHP), so that aligns with our terminology.  The "secure" word also hints to users that this is not related to dev productivity (such as Edit-and-Continue scenarios).

aganea commented:
> What would you think of renaming the options to `-fms-secure-hotpatch-functions-file` and `-fms-secure-hotpatch-functions-list`? The feature internally is known as "Secure Hot-Patching" (SHP), so that aligns with our terminology. The "secure" word also hints to users that this is not related to dev productivity (such as Edit-and-Continue scenarios).

The renaming sounds good. Are you able to refer to this feature everywhere in LLVM documentation or comments, or commits, as "Secure Hot-Patching" in that case?

In addition to a more elaborate description, can you also please add a web link in `WindowsHotPatch.cpp` linking to the most descriptive webpage about this feature?

aganea commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -453,6 +453,37 @@ CodeGenModule::CodeGenModule(ASTContext &C,
+      }
+    }
+
+    for (const auto &FuncName : CGO.MSHotPatchFunctionsList) {
```
LLVM coding standards [instructs](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements) to omit braces for one liners, but TBH I prefer the Rust rule where the brace should always be mandatory. What do other reviewers think?

aganea replied:
I'd say let's stick to the LLVM guidelines and remove the extra braces as suggested in https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

aganea commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -453,6 +453,37 @@ CodeGenModule::CodeGenModule(ASTContext &C,
+      this->MSHotPatchFunctions.push_back(FuncName);
+    }
+
+    std::sort(this->MSHotPatchFunctions.begin(),
```
You can just write `llvm::sort(this->MSHotPatchFunctions);`.

aganea commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -6946,6 +6946,16 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
 
   Args.AddLastArg(CmdArgs, options::OPT_fms_hotpatch);
 
+  if (Arg *A = Args.getLastArg(options::OPT_fms_hotpatch_functions_file)) {
```
Idem, extra brace.

aganea commented on clang/test/CodeGen/ms-hotpatch-lto.c:
```diff
@@ -0,0 +1,18 @@
+// This verifies that hotpatch function attributes are correctly propagated through LLVM IR when compiling with LTO.
+//
+// RUN: %clang_cl -c --target=x86_64-windows-msvc -O2 /Z7 -fms-hotpatch-functions-file=%S/ms-hotpatch-functions.txt -flto /Fo%t.bc %s
```
If you wish, you could also you `split-file` before this line to keep the test self-contained. See https://llvm.org/docs/TestingGuide.html#extra-files

sivadeilra replied:
Good to know.  I've changed this to a single `echo`, which I think is simpler than using `split-file`.  If it gets any more complicated than that, I'll use `split-file`.

aganea commented on llvm/lib/CodeGen/AsmPrinter/CodeViewDebug.cpp:
```diff
@@ -807,6 +809,32 @@ void CodeViewDebug::emitObjName() {
   endSymbolRecord(CompilerEnd);
 }
 
+void CodeViewDebug::emitHotPatchInformation() {
```
As mentionned before, maybe better to rename to `emitSecureHotPatchInformation()`

sivadeilra replied:
> Would that tooling remain proprietary, or is it something that will be open-sourced as well, for example for NT drivers developpers?

In the near term, there are no plans on releasing these tools.  The only scenarios for these tools require signing using the same certs that we use for shipping OS components, so even if we provided the tools, there isn't currently a scenario where they could be used.

I could imagine that changing in the future, such as for device drivers, but I'm not in a position to speak for that direction.

aganea commented on llvm/lib/CodeGen/CMakeLists.txt:
```diff
@@ -250,6 +250,7 @@ add_llvm_component_library(LLVMCodeGen
   VirtRegMap.cpp
   WasmEHPrepare.cpp
   WindowScheduler.cpp
+  WindowsHotPatch.cpp
```
Similarily, `WindowsSecureHotPatch.cpp`?

sivadeilra replied:
Renamed to `WindowsSecureHotPatching.cpp`

aganea commented on llvm/lib/CodeGen/WindowsHotPatch.cpp:
```diff
@@ -0,0 +1,242 @@
+        continue;
+      }
+
+      if (HotPatchFunctionsSet.contains(F.getName())) {
```
Extra braces.

sivadeilra replied:
removed

aganea commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -2649,6 +2649,15 @@ void CodeGenModule::ConstructAttributeList(StringRef Name,
+    if (!MSHotPatchFunctions.empty()) {
+      bool IsHotPatched = std::binary_search(MSHotPatchFunctions.begin(),
+                                             MSHotPatchFunctions.end(), Name);
+      if (IsHotPatched) {
```
Braces.

sivadeilra replied:
Thanks.  I've done a full pass through the patch now, and I believe I've fixed all of the use of braces.

aganea commented on llvm/lib/CodeGen/WindowsSecureHotPatching.cpp:
```diff
@@ -0,0 +1,294 @@
+
+using namespace llvm;
+
+#define DEBUG_TYPE "windows-hot-patch"
```
`"windows-secure-hot-patch"` ?

aganea commented:
Thank you for all the changes. Do you mind also including "secure" in the commit title?

aganea reviewed (commented):
The patch looks good to me, but I have one last question:

sivadeilra commented on llvm/lib/CodeGen/WindowsSecureHotPatching.cpp:
```diff
@@ -0,0 +1,283 @@
+        continue;
+
+      if (HotPatchFunctionsSet.contains(F.getName()))
+        F.addFnAttr(Attribute::MarkedForWindowsHotPatching);
```
> What is the plan there for handling the function prologue [...] Shouldn't we enable `-fms-hotpatch` de facto for the TU when using the `-fms-secure-hotpatch` ... flags?

It's true that `-fms-secure-*` requires `-fms-hotpatch`.  We could: 1) do nothing in the compiler (and detect it downstream during patch generation); 2) report a warning when `-fms-secure-*` is used but `-fms-hotpatch` is not; 3) implicitly turn on `-fms-hotpatch` when `-fms-secure-*` is used.

I don't have a strong preference, here.  In Windows, we compile everything with `/hotpatch`, because the codegen costs are minor and because this maximizes the set of code that we can hot-patch down the road.  I'm willing to go with any of the above options -- just let me know which one the LLVM project would prefer.

> What is the plan for [...] leaving space for the trampoline, when using lld-link?

Our current plans are to continue using the MSVC linker. There are a lot of behaviors in the MSVC linker (unrelated to hot-patching) that we depend on; migrating to LLD would be a huge task.  However, if we did go down the path of enabling SHP using LLD, the main requirement is to support the `/functionpadmin` argument.  Fortunately, that option is publicly documented and its behavior is relatively simple, so implementing that in LLD would be straight-forward.

The last piece of the puzzle is constructing the final binary image by editing in the `.hpdata` COFF section.  That's fairly easy, too.

> What about cumulative hot patches?

This is already supported and used extensively.  Each patch is a complete binary image on its own (contains all of the code and data needed to run standalone).  However, each patch also describes the base image that it patches.  The base image may be an original image (which contains no hot-patches) or may be another hot-patch image.

Whenever a hot-patch is applied, it modifies the code in the original base image to point to patched functions.  In patched images, when a function A calls B, we know the caller A is patched (or else we wouldn't be in this module).  If B is also hot-patched, then A will directly call B.  However, if B was _not_ hot-patched, then B is patched to redirect to the version of B in the base image.  This means that patched-A will call original-B, even though the hot-patch image for A has a complete copy of B.

We do this so that the "canonical" address for B is in the original image.  If a second hot-patch is applied, then B might be patched.  So the B in the original image would be rewritten to point to the B in the second hot-patch.  If a thread was already running code in patched-A but has not yet started calling B, then patch1-A will call patch2-B, which is what we want to happen.  Then, once patch1-A returns, then the next call to A will enter patch2-A.  So while we can't evict calls to functions that are currently running, we can redirect them to new code as soon as they return.

aganea replied:
> > What is the plan there for handling the function prologue [...] Shouldn't we enable `-fms-hotpatch` de facto for the TU when using the `-fms-secure-hotpatch` ... flags?
> 
> It's true that `-fms-secure-*` requires `-fms-hotpatch`. We could: 1) do nothing in the compiler (and detect it downstream during patch generation); 2) report a warning when `-fms-secure-*` is used but `-fms-hotpatch` is not; 3) implicitly turn on `-fms-hotpatch` when `-fms-secure-*` is used.
> 
> I don't have a strong preference, here. In Windows, we compile everything with `/hotpatch`, because the codegen costs are minor and because this maximizes the set of code that we can hot-patch down the road. I'm willing to go with any of the above options -- just let me know which one the LLVM project would prefer.

Ok, fair enough, since you already have that enabled in your build scripts that use the `-fms-secure-hotpatch` flags, and since these flags are for a very limited audience, we could leave this as it is for now.

> the main requirement is to support the `/functionpadmin` argument.

This is [already supported](https://reviews.llvm.org/D49366) in LLD.

aganea reviewed (approved):
LGTM. Perhaps you might want to wait a few days before commiting, in case others would like to comment.

sivadeilra commented:
@aganea , I've encountered one problem with this PR.  The pass which creates the `__ref_*` variables needs to run before `CodeViewDebug::collectGlobalVariableInfo`, but the way we did this in this PR breaks several other integration tests.  From looking at the sequence of passes, `CodeViewDebug::collectGlobalVariableInfo` runs under `AsmPrinter::doInitialization`.

What's the best place to do this `__ref_*` rewrite pass?

dpaoliello commented:
@efriedma-quic @mstorsjo any objections to merging?

mstorsjo commented:
No objections from me. (I haven’t had time to look at it in detail, but it doesn’t seem like an area I’m familiar with anyway, and it sounds like @aganea has given it a thorough check.)

efriedma-quic reviewed (commented):
It looks like my previous comments about ConstantExpr never got resolved?  Specifically, what happens if you have an instruction where an operand is a ConstantExpr GEP of a global, or something like that.

efriedma-quic reviewed (commented):
Also, the no-op ConstantExpr::getGetElementPtr call is still there.

sivadeilra commented:
@efriedma-quic Thanks, I've confirmed that removing the GEP results in the same output code.  I'll make that change in the next commit.

Could you help me understand a situation where a global would be a ConstantExpr?  Would it be something like:

```
int g_foo;
int g_ptr = &g_foo;
```

sivadeilra commented:
Thanks for the example.  Yeah, that's not working correctly.  I'll work on this.

sivadeilra commented:
@efriedma-quic - I'm looking at several approaches for handling this, but before I go too far with it, it would help to know just how general or complete the solution needs to be.  The problem is that the transformation I'm doing will necessarily change a constant expression tree into a non-constant expression tree.  So every time I find a reference to a global variable that is inside a constant expression, I will need to rewrite it to point to the load instruction, and rewrite its caller (and that caller...) until I reach a non-constant user.

I'm thinking that the way to do this is to iterate through instructions, then recursively walk every operand and look for `GlobalVariable` references.  If I find any `ConstantExpr` operands, then I'll recursively descend into that `ConstantExpr` and look for `GlobalVariableVal` and `GetElementPtr` constant expressions.

My questions for you:
* Is rewriting these constant expressions trees sufficient?
* Will the results be well-formed?
* How many different cases will this need to cover?

Our hot-patching tools disassemble instruction streams and look for references to global variables that the compiler did not transform into `__ref_*` indirections.  So if we don't match some particular case, we should still catch it and reject the patch.  Rejecting the patch is better than generating invalid code, of course.

In this first PR, I'm mainly looking for getting the `S_HOTPATCHFUNC` generation in place, and covering the 95% cases, which access an ordinary global variable.  Handing a field dereference or an array dereference, both of which should show up as GEPs, should be sufficient, right?

efriedma-quic commented:
You might want to look at convertUsersOfConstantsToInstructions, which addresses similar use-cases.  (You might be able to use it directly?  If not, you can at least reference the implementation.)

Recursively rewriting expression trees should work, and result in well-defined Instructions.  We used to have constants with weird constraints, but those are gone in current LLVM.

For "normal" code, you should only run into ConstantExpr and subclasses of ConstantAggregate.  There are a few other kinds of constant expressions that could show up with specific flags or unusual constructs (ConstantPtrAuth, NoCFIValue, DSOLocalEquivalent, BlockAddress).

nikic commented on llvm/lib/CodeGen/WindowsSecureHotPatching.cpp:
```diff
@@ -0,0 +1,287 @@
+
+    if (!LLVMMSSecureHotPatchFunctionsFile.empty()) {
+      auto BufOrErr =
+          llvm::MemoryBuffer::getFile(LLVMMSSecureHotPatchFunctionsFile);
```
```suggestion
          MemoryBuffer::getFile(LLVMMSSecureHotPatchFunctionsFile);
```
nit: Do not use `llvm::` qualifiers in `llvm/`.

sivadeilra commented:
An update: I need to be away from this work for several days, but will return to it on June 3.  I will update this with the changes needed for the deeper inspection / rewriting of const expressions when I do.

Thanks.

sivadeilra commented:
Ok, I've redone the code that modifies GlobalVariable references.  It is more complicated, but that was necessary to handle constant expressions.  There are detailed comments in the code, so I'll let those comments speak for themselves.

Fixed a bug in how the `__ref_*` variables were created.  They were being created as constants, but that's incorrect; it allows the optimizer to read the initializer / definition and then inline it into places that use the GV.  That defeats the whole purpose of the redirection mechanism.  I've changed the variables so that they are non-constant.

Fixed a bug where the `__ref_*` variables were marked for internal linkage.  They need to be external-but-weak.

dpaoliello commented on llvm/include/llvm/Bitcode/LLVMBitCodes.h:
```diff
@@ -798,6 +798,8 @@ enum AttributeKindCodes {
   ATTR_KIND_SANITIZE_TYPE = 101,
   ATTR_KIND_CAPTURES = 102,
+  ATTR_KIND_ALLOW_DIRECT_ACCESS_IN_HOT_PATCH_FUNCTION = 103,
+  ATTR_KIND_MARKED_FOR_WINDOWS_HOT_PATCHING = 104,
```
For consistency

```suggestion
  ATTR_KIND_MARKED_FOR_WINDOWS_SECURE_HOT_PATCHING = 104,
```

dpaoliello commented on llvm/include/llvm/IR/Attributes.td:
```diff
@@ -389,6 +389,17 @@ def CoroDestroyOnlyWhenComplete : EnumAttr<"coro_only_destroy_when_complete", In
 def CoroElideSafe : EnumAttr<"coro_elide_safe", IntersectPreserve, [FnAttr]>;
 
+/// Function is marked for Windows Hot Patching
+def MarkedForWindowsHotPatching
```
Same here, for consistency, "secure" hot patching

dpaoliello commented on llvm/lib/CodeGen/WindowsSecureHotPatching.cpp:
```diff
@@ -0,0 +1,599 @@
+// that all code (both unpatched and patched) is using the same instances of
+// global variables.
+//
+// The Windows hot-patch infrastructure handles initializing these __ref_*
```
Should this be "modifying" or "redirecting" instead of "initializing"? As in, then point to the GVs in the current image but Windows will change them to point to the base image if required.

sivadeilra replied:
Changed to "modifying"

dpaoliello commented on llvm/lib/CodeGen/WindowsSecureHotPatching.cpp:
```diff
@@ -0,0 +1,599 @@
+  return NewInst;
+}
+
+// Processes a function that is marked for hot-patching.
```
This comment seems out of place

sivadeilra replied:
Moved this comment (and several others)

dpaoliello commented on llvm/lib/CodeGen/WindowsSecureHotPatching.cpp:
```diff
@@ -0,0 +1,599 @@
+    for (unsigned OpIndex = 0; OpIndex < NumOperands; ++OpIndex) {
+      Value *Op = U->getOperand(OpIndex);
+      // Do not use short-circuiting, here. We need to traverse the whole tree.
+      FoundAny |= searchConstantExprForGlobalVariables(Op, GVLoadMap, GVUses);
```
Would it make sense to keep track of constants we've walked previously to avoid walking them again?

sivadeilra replied:
I don't think so.  We already keep track of instructions.  The constant expressions are generally quite small, and we're only processing them in functions that are marked for hot-patching to begin with.

dpaoliello reviewed (commented):
Can you please add tests for the constant expression rewriting stuff, and not redirecting constant GVs without pointers?

sivadeilra commented:
> Can you please add tests for the constant expression rewriting stuff, and not redirecting constant GVs without pointers?

It already has some, in `ms-secure-hotpatch-globals.c`.

efriedma-quic commented on llvm/lib/CodeGen/WindowsSecureHotPatching.cpp:
```diff
@@ -0,0 +1,596 @@
+
+  SmallVector<Value *, 8> ReplacedOperands;
+
+  if (V->getValueID() == Value::GlobalVariableVal) {
```
```
if (auto *GV = dyn_cast<GlobalVariable>(V)) {
```

efriedma-quic commented on llvm/lib/CodeGen/WindowsSecureHotPatching.cpp:
```diff
@@ -0,0 +1,596 @@
+}
+
+static bool searchConstantExprForGlobalVariables(
+    Value *V, SmallDenseMap<GlobalVariable *, Value *> &GVLoadMap,
```
I think the first argument to searchConstantExprForGlobalVariables should always be a Constant?

efriedma-quic commented on llvm/lib/CodeGen/WindowsSecureHotPatching.cpp:
```diff
@@ -0,0 +1,596 @@
+    }
+  }
+
+  if (User *U = dyn_cast<User>(V)) {
```
You probably want to handle other `GlobalValue`s.

sivadeilra replied:
I don't see any other subclasses of GlobalValue that could contain references to GlobalVariables.  Can you provide an example?

efriedma-quic reviewed (commented):
I'd like to see regression tests for the cases I've mentioned.  PHI nodes, ConstantExpr, and I guess the combination of the two.

efriedma-quic commented:
For the constant expression handling stuff, I think you have the basic framework working correctly; if we run into additional issues, I'm happy to leave handling them for followups.

sivadeilra commented on llvm/include/llvm/IR/Attributes.td:
```diff
@@ -389,6 +389,17 @@ def CoroDestroyOnlyWhenComplete : EnumAttr<"coro_only_destroy_when_complete", In
+/// decl, not the hotpatched function.
+def AllowDirectAccessInHotPatchFunction
+    : EnumAttr<"allow_direct_access_in_hot_patch_function",
+               IntersectPreserve, [FnAttr]>;
```
Sure, I'm happy to make that change.  Since these attributes are used very, very rarely, it makes sense not to burn up bits that could be used for something better.  Is there an example I can use as a model?

sivadeilra commented on llvm/test/CodeGen/Generic/ms-secure-hotpatch-attr.ll:
```diff
@@ -0,0 +1,38 @@
+; This tests directly annotating a function with marked_for_windows_hot_patching.
+;
+; RUN: llc -mtriple=x86_64-windows < %s | FileCheck %s
```
Done

efriedma-quic commented on clang/test/CodeGen/X86/ms-secure-hotpatch-cpp.cpp:
```diff
@@ -0,0 +1,22 @@
+// and that name mangling works as expected.
+//
+// RUN: %clang_cl -c --target=x86_64-windows-msvc -O2 /Z7 -fms-secure-hotpatch-functions-list=?this_gets_hotpatched@@YAHXZ /Fo%t.obj %s
+// RUN: llvm-readobj --codeview %t.obj | FileCheck %s
```
I checked to confirm, and I guess clang/test/CodeGen/X86 doesn't have a implicit lit.cfg file, so you probably need `REQUIRES: x86-registered-target` for all the clang tests.  (The llvm/ tests are fine.)

efriedma-quic reviewed (approved):
LGTM after the REQUIRES issue is addressed.

efriedma-quic commented:
Linux CI failure is probably #145407.  Given that, I'm fine with just ignoring it, instead of trying to rerun.

For reference, the only way to force a rerun is to push a new patch.  (It doesn't matter whether it contains any significant changes, as long as the commit hash is different.)

sivadeilra commented:
@efriedma-quic Thank you for working with me on this PR!

qinkunbao commented:
Hi, can you take a look at the buildbot failures.

https://lab.llvm.org/buildbot/#/builders/51/builds/18489
